cmd/archeio/main_test.go

```diff
@@ -17,121 +20,99 @@ limitations under the License.
 package main
 
 import (
+	"context"
 	"net/http"
-	"net/http/httptest"
+	"os"
+	"os/exec"
 	"testing"
-)
+	"time"
 
-var (
-	defaultUpstreamRegistry = "https://k8s.gcr.io"
+	"sigs.k8s.io/oci-proxy/internal/integration"
 )
 
-type request struct {
-	path     string
-	redirect bool
-}
+// TestIntegrationMain tests the entire, built binary with an integration
+// test, pulling images with crane
+func TestIntegrationMain(t *testing.T) {
+	integration.MaybeSkip(t)
 
-type expected struct {
-	url        string
-	path       string
-	statusCode int
-}
+	// setup crane
+	rootDir, err := integration.ModuleRootDir()
+	if err != nil {
+		t.Fatalf("Failed to detect module root dir: %v", err)
+	}
+	// NOTE: also ensures rootDir/bin is in front of $PATH
+	if err := integration.EnsureCrane(rootDir); err != nil {
+		t.Fatalf("Failed to ensure crane: %v", err)
+	}
 
-type scenario struct {
-	name     string
-	request  request
-	expected expected
-}
+	// build binary
+	buildCmd := exec.Command("make", "archeio")
+	buildCmd.Dir = rootDir
+	if err := buildCmd.Run(); err != nil {
+		t.Fatalf("Failed to build archeio for integration testing: %v", err)
+	}
 
-type suite struct {
-	handler   http.Handler
-	scenarios []scenario
-	tests     []func(resp *http.Response, scenario scenario)
-}
+	// start server in background
+	testPort := "61337"
```

Comment:
why not pick a random one with :0?

Reply:
:effort:
If I pick it by binding a port in the test I need to make sure to bind to the same addr and hope the port doesn't get used between releasing it and starting the server anyhow.
If I let the server process pick it then I need to discover it.

Reply:
it is highly unlikely that happens, or we should have the CI failing constantly , ... not a big deal though

Reply:
CI doesn't have peers opening high ports because we run tests in a network namespace for those tests. Local dev may.

Reply:
you can't run the test continously, `-count 2` the port closed has a default TIME_WAIT status , 1min IIRC , it should not be able to bind it

Reply:
> you can't run the test continously, -count 2 the port closed has a default TIME_WAIT status , 1min IIRC , it should not be able to bind it

That's not something you should be doing anyhow, this is a full-on integration test, we have unit tests in `cmd/archeio/app` testing the http handlers etc directly you can run continuously, the test here is just to exercise the binary. it's more like an e2e.

Reply:
+1 to merge and iterate